cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -72,32 +99,48 @@ func (pe *ProbeEvent) addBlockDeviceEvent(msg controller.EventMessage) {
 
 // deleteBlockDeviceEvent deactivate blockdevice resource using uuid from etcd
 func (pe *ProbeEvent) deleteBlockDeviceEvent(msg controller.EventMessage) {
-	blockDeviceOk := pe.deleteBlockDevice(msg)
-
-	// when one disk is removed from node and entry related to
-	// that disk is not present in etcd,  in that case it
-	// again rescan full system and update etcd accordingly.
-	if !blockDeviceOk {
-		go pe.initOrErrorEvent()
-	}
-}
-
-func (pe *ProbeEvent) deleteBlockDevice(msg controller.EventMessage) bool {
-	bdList, err := pe.Controller.ListBlockDeviceResource()
+	bdAPIList, err := pe.Controller.ListBlockDeviceResource(false)
 	if err != nil {
 		klog.Error(err)
-		return false
 	}
-	ok := true
-	for _, diskDetails := range msg.Devices {
-		existingBlockDeviceResource := pe.Controller.GetExistingBlockDeviceResource(bdList, diskDetails.UUID)
+
+	isDeactivated := true
+	isGPTBasedUUIDEnabled := pe.Controller.FeatureGates.IsEnabled(features.GPTBasedUUID)
+
+	for _, device := range msg.Devices {
+		// create the new UUID for removing the device
+		if isGPTBasedUUIDEnabled {
+			_, ok := pe.Controller.BDHierarchy[device.DevPath]
+			if !ok {
+				klog.Infof("Disk %s not in hierarchy", device.DevPath)
+				// not in hierarchy continue
+				continue
+			}
+			// remove from the hierarchy
+			delete(pe.Controller.BDHierarchy, device.DevPath)
+
+			uuid, ok := generateUUID(*device)
+			if !ok {
+				klog.Info("could not recreate UUID while removing device")
```

Comment:
May be a comment on when such cases will occur will be good. 

This case can probably occur if a disk was being used for cstor (claimed) prior to upgrading to new uuid changes. And the disk is now removed from the node.

Reply:
Added.

No @kmova . This is the case when a remove event is generated for a device without any unique identifier. say a disk without WWN, we will just ignore it as there is no such blockdevice.

NOTE: This PR does not handle any cases of upgrade. It has to be done separately. All changes here are with the assumption this is a fresh installation.